File: Duomenų struktūrų naudojimas/C++/Užduotis 20241007_01_sprendimas_(naudojimas_steku_deklu_eile).cpp
#include <iostream>
#include <queue>
#include <stack>
#include <deque>
#include <string>  // For using std::string

// Using standard library types without std:: prefix
using namespace std;

class BankTransactions {
public:
    queue<string> transactionQueue;  // Naudojame kaip eilę
    stack<string> transactionStack;  // Naudojame kaip steką
    deque<string> transactionDeque;  // Naudojame kaip deklą

    // Eilės funkcijos
    void enqueue_transaction(const string& transaction) {
        transactionQueue.push(transaction);
    }

    string dequeue_transaction() {
        if (!transactionQueue.empty()) {
            string front = transactionQueue.front();
            transactionQueue.pop();
            return front;
        }
        return "Queue is empty";
    }

    // Steko funkcijos
    void push_transaction(const string& transaction) {
        transactionStack.push(transaction);
    }

    string pop_transaction() {
        if (!transactionStack.empty()) {
            string top = transactionStack.top();
            transactionStack.pop();
            return top;
        }
        return "Stack is empty";
    }

    // Deklo funkcijos
    void add_front_transaction(const string& transaction) {
        transactionDeque.push_front(transaction);
    }

    void add_rear_transaction(const string& transaction) {
        transactionDeque.push_back(transaction);
    }

    string remove_front_transaction() {
        if (!transactionDeque.empty()) {
            string front = transactionDeque.front();
            transactionDeque.pop_front();
            return front;
        }
        return "Deque is empty";
    }

    string remove_rear_transaction() {
        if (!transactionDeque.empty()) {
            string back = transactionDeque.back();
            transactionDeque.pop_back();
            return back;
        }
        return "Deque is empty";
    }
};

int main() {
    BankTransactions bt;

    // Eilės operacijos
    bt.enqueue_transaction("Client 1");
    bt.enqueue_transaction("Client 2");
    cout << "Eilėje aptarnaujamas: " << bt.dequeue_transaction() << endl;  // Client 1

    // Steko operacijos
    bt.push_transaction("Client 1");
    bt.push_transaction("Client 2");
    cout << "Steke aptarnaujamas: " << bt.pop_transaction() << endl;  // Client 2

    // Deklo operacijos
    bt.add_rear_transaction("Client 1");
    bt.add_front_transaction("VIP Client");
    cout << "Dekle aptarnaujamas iš priekio: " << bt.remove_front_transaction() << endl;  // VIP Client
    cout << "Dekle aptarnaujamas iš galo: " << bt.remove_rear_transaction() << endl;  // Client 1

    return 0;
}

File: Duomenų struktūrų naudojimas/Python/Užduotis 20241007_01_sprendimas_(naudojimas_steku_deklu_eile).py
from collections import deque

# Banko operacijų klasė
class BankTransactions:
    def __init__(self):
        self.queue = deque()  # Naudojame kaip eilę
        self.stack = deque()  # Naudojame kaip steką
        self.deque = deque()  # Naudojame kaip deklą
    
    # Eilės funkcijos
    def enqueue_transaction(self, transaction):
        self.queue.append(transaction)
        print(f"[Elementas {transaction} įdėti į eilę]")
    
    def dequeue_transaction(self):
        if self.queue:
            print("[Gauname elementą iš eilės]")
            return self.queue.popleft()
        return None
    
    # Steko funkcijos
    def push_transaction(self, transaction):
        self.stack.append(transaction)
    
    def pop_transaction(self):
        if self.stack:
            return self.stack.pop()
        return None
    
    # Deklo funkcijos
    def add_front_transaction(self, transaction):
        self.deque.appendleft(transaction)
    
    def add_rear_transaction(self, transaction):
        self.deque.append(transaction)
    
    def remove_front_transaction(self):
        if self.deque:
            return self.deque.popleft()
        return None
    
    def remove_rear_transaction(self):
        if self.deque:
            return self.deque.pop()
        return None

# Pavyzdys
bt = BankTransactions()

# Eilės operacijos
bt.enqueue_transaction("Client 1")
bt.enqueue_transaction("Client 2")
print("Eilėje aptarnaujamas:", bt.dequeue_transaction())  # Client 1

# Steko operacijos
bt.push_transaction("Client 1")
bt.push_transaction("Client 2")
print("Steke aptarnaujamas:", bt.pop_transaction())  # Client 2

# Deklo operacijos
bt.add_rear_transaction("Client 1")
bt.add_front_transaction("VIP Client")
print("Dekle aptarnaujamas iš priekio:", bt.remove_front_transaction())  # VIP Client
print("Dekle aptarnaujamas iš galo:", bt.remove_rear_transaction())  # Client 1
